File: mobile/modules/readium/ios/BookService.swift

import Combine
import Foundation
import ReadiumShared
import ReadiumStreamer
import ReadiumFuzi
import ZIPFoundation
import OrderedCollections

enum BookServiceError : Error {
    case unopenedPublication(String)
    case locatorNotInReadingOrder(String, String)
    case noMediaOverlay(String, String)
    case openContent(String, String)
    case extractFailed(URL, String)
}

extension Task where Failure == Error {
    /// Performs an async task in a sync context.
    ///
    /// - Note: This function blocks the thread until the given operation is finished. The caller is responsible for managing multithreading.
    static func synchronous(priority: TaskPriority? = nil, operation: @escaping @Sendable () async throws -> Success) {
        let semaphore = DispatchSemaphore(value: 0)

        Task(priority: priority) {
            defer { semaphore.signal() }
            return try await operation()
        }

        semaphore.wait()
    }
}

final class BookService {
    let retriever: AssetRetriever
    private let opener: PublicationOpener
    private var publications: [String : Publication] = [:]
    private var clips: [String : [OverlayPar]] = [:]

    public static let shared = BookService()

    private static func onCreatePublication(_ manifest: inout Manifest,_ container: inout Container,_ services: inout PublicationServicesBuilder) async -> Void {
        let localContainer = container

        let opfHREF = try? await EPUBContainerParser(container: localContainer).parseOPFHREF()

        guard let opfHREF = opfHREF else {
            return
        }

        guard let opfResource = localContainer[opfHREF] else {
            return
        }

        guard let opfData = try? await opfResource.read().get() else {
            return
        }

        guard let document = try? ReadiumFuzi.XMLDocument(data: opfData) else {
            return
        }
        document.definePrefix("opf", forNamespace: "http://www.idpf.org/2007/opf")

        let manifestItems = document.xpath("/opf:package/opf:manifest/opf:item")

        var readingOrder = manifest.readingOrder
        for manifestItem in manifestItems {
            guard let relativeHREF = manifestItem.attr("href").flatMap(RelativeURL.init(epubHREF:)) else {
                continue
            }

            guard let href = opfHREF.resolve(relativeHREF) else {
                continue
            }

            guard let readingOrderIndex = readingOrder.firstIndexWithHREF(href) else {
                continue
            }

            var link = readingOrder[readingOrderIndex]

            if let mediaOverlayId = manifestItem.attr("media-overlay"),
               let mediaOverlay = manifestItems.first(where: { $0.attr("id") == mediaOverlayId }),
               let relativeMediaOverlayHref = mediaOverlay.attr("href").flatMap(RelativeURL.init(epubHREF:)),
               let mediaOverlayHref = opfHREF.resolve(relativeMediaOverlayHref) {
                link.addProperties(["mediaOverlay": mediaOverlayHref.string])
                readingOrder[readingOrderIndex] = link
            }
        }

        manifest.readingOrder = readingOrder
    }

    private init() {
        retriever = AssetRetriever(
            formatSniffer: DefaultFormatSniffer(),
            resourceFactory: DefaultResourceFactory(httpClient: DefaultHTTPClient()),
            archiveOpener: DefaultArchiveOpener()
        )

        opener = PublicationOpener(parser: EPUBParser())
    }

    func extractArchive(archiveUrl: URL, extractedUrl: URL) throws {

        let fileManager = FileManager.default

        do {
            try fileManager.createDirectory(at: extractedUrl, withIntermediateDirectories: true, attributes: nil)
            try fileManager.unzipItem(at: archiveUrl, to: extractedUrl)
        } catch {
            print(error.localizedDescription)
            throw BookServiceError.extractFailed(archiveUrl, error.localizedDescription)
        }
    }

    func getPublication(for bookId: String) -> Publication? {
        return publications[bookId]
    }

    func getResource(for bookId: String, link: Link) throws -> Resource {
        guard let publication = publications[bookId] else {
            throw BookServiceError.unopenedPublication(bookId)
        }
        return publication.get(link)!
    }

    func getPositions(for bookId: String) async throws -> [Locator] {
        guard let publication = getPublication(for: bookId) else {
            throw BookServiceError.unopenedPublication(bookId)
        }
        return await publication.positions().getOrNil() ?? []
    }
    
    func getOverlayClips(for bookUuid: String) -> [OverlayPar] {
        return clips[bookUuid] ?? []
    }

    func getClip(for bookId: String, locator: Locator) -> OverlayPar? {
        guard let bookClips = clips[bookId] else {
            return nil
        }
        
        guard let fragment = locator.locations.fragments.first else {
            return nil
        }
        
        return bookClips.first {
            $0.locator.href.string == locator.href.string &&
            $0.fragmentId == fragment
        }
    }

    func locateFromPositions(for bookId: String, link: Link) async throws -> Locator {
        guard let publication = getPublication(for: bookId) else {
            throw BookServiceError.unopenedPublication(bookId)
        }

        guard let readingOrderIndex = publication.readingOrder.firstIndexWithHREF(RelativeURL(epubHREF: link.href)!) else {
            throw BookServiceError.locatorNotInReadingOrder(bookId, link.href)
        }

        let positions = try await publication.positionsByReadingOrder().get()

        guard let locator = positions[readingOrderIndex].first else {
            throw BookServiceError.locatorNotInReadingOrder(bookId, link.href)
        }

        return locator
    }

    func getFragments(for bookId: String, locator: Locator) -> [OverlayPar] {
        guard let bookClips = clips[bookId] else {
            return []
        }
        return bookClips.filter { $0.locator.href.string == locator.href.string }
    }

    func getFragment(for bookId: String, before locator: Locator) -> OverlayPar? {
        guard let currentFragment = locator.locations.fragments.first else {
            return nil
        }
        guard let bookClips = clips[bookId] else {
            return nil
        }
        guard let currentIndex = bookClips.firstIndex(where: { $0.locator.href.string == locator.href.string && $0.fragmentId == currentFragment }) else {
            return nil
        }
        if currentIndex == bookClips.startIndex {
            return nil
        }
        let previousIndex = bookClips.index(before: currentIndex)
        return bookClips[previousIndex]
    }

    func getFragment(for bookId: String, after locator: Locator) -> OverlayPar? {
        guard let currentFragment = locator.locations.fragments.first else {
            return nil
        }
        guard let bookClips = clips[bookId] else {
            return nil
        }
        guard let currentIndex = bookClips.firstIndex(where: { $0.locator.href.string == locator.href.string && $0.fragmentId == currentFragment }) else {
            return nil
        }
        if currentIndex == bookClips.endIndex {
            return nil
        }
        let nextIndex = bookClips.index(after: currentIndex)
        return bookClips[nextIndex]
    }

    func getLocatorFor(bookId: String, href: String, fragment: String) -> Locator? {
        guard let bookClips = clips[bookId] else {
            return nil
        }
        
        return bookClips.first { $0.locator.href.string == href && $0.fragmentId == fragment }?.locator
    }

    func getFragment(for bookId: String, clipUrl: URL, position: Double) -> OverlayPar? {
        guard let bookClips = clips[bookId] else {
            return nil
        }
        
        let clipsInUrl = bookClips.filter { $0.relativeUrl.relativeString == clipUrl.relativeString }
        
        return searchForClip(clips: clipsInUrl, position: position)
    }

    func locateLink(for bookId: String, link: Link) async -> Locator? {
        guard let publication = getPublication(for: bookId) else {
            return nil
        }
        return await publication.locate(link)
    }

    private func makeMediaOverlays(for bookId: String, pub publication: Publication) async throws {
        var bookClips: [OverlayPar] = []

        let mediaOverlayLinks = publication.resources.filterByMediaType(.smil)

        for link in mediaOverlayLinks {
            let mediaOverlays = STMediaOverlays(link: link)
            let node = STMediaOverlayNode()

            let smilResource = publication.get(link)!
            guard let smilData = try? await smilResource.read().get(),
                let smilXml = try? XMLDocument(data: smilData) else
            {
                throw OPFParserError.invalidSmilResource
            }

            smilXml.definePrefix("smil", forNamespace: "http://www.w3.org/ns/SMIL")
            smilXml.definePrefix("epub", forNamespace: "http://www.idpf.org/2007/ops")
            guard let body = smilXml.firstChild(xpath: "./smil:body") else {
                throw OPFParserError.invalidSmilResource
            }

            node.role.append("section")
            if let textRef = body.attr("textref") { // Prevent the crash on the japanese book
                node.text = RelativeURL(epubHREF: link.href)!.resolve( RelativeURL(epubHREF: textRef)!)!.string
            }

            // get body parameters <par>a
            await STSMILParser.parseParallels(publication, in: body, withParent: node, base: link.href, htmlContent: nil)
            await STSMILParser.parseSequences(publication, in: body, withParent: node, mediaOverlays: mediaOverlays, base: link.href)

            bookClips.append(contentsOf: mediaOverlays.clips())
        }

        clips[bookId] = bookClips
    }

    /// Opens the Readium 2 Publication at the given `url`.
    func openPublication(for bookId: String, at url: FileURL, clips: [OverlayPar]?) async throws -> Publication {
        if let publication = getPublication(for: bookId) {
            return publication
        }

        let container = try! await DirectoryContainer(directory: url)

        let asset = await retriever.retrieve(container: container).getOrNil()

        let pubResult = await opener.open(asset: asset!, allowUserInteraction: false, onCreatePublication: BookService.onCreatePublication)

        do {
            let pub = try pubResult.get()
            try checkIsReadable(publication: pub)
            publications[bookId] = pub
            if clips == nil {
                try await makeMediaOverlays(for: bookId, pub: pub)
            } else {
                self.clips[bookId] = clips
            }
            return pub
        } catch {
            throw BookError.openFailed(error)
        }
    }

    func buildAudiobookManifest(for bookId: String) async throws -> Manifest {
        guard let publication = getPublication(for: bookId) else {
            throw BookError.bookNotFound
        }
        guard let bookClips = clips[bookId] else {
            throw BookError.bookNotFound
        }
        let clipsByHref = Dictionary(grouping: bookClips, by: \.locator.href.string)

        func buildAudiobookTocLink(link: Link) async throws -> Link? {
            let children = try await link.children.asyncMap {
                try await buildAudiobookTocLink(link: $0)
            }.compactMap { $0 }

            let fallbackLink = children.first.map {
                Link(
                    href: $0.href,
                    mediaType: $0.mediaType,
                    title: link.title,
                    duration: 0,
                    children: children,
                )
            }

            guard let tocLocator = await locateLink(for: bookId, link: link) else {
                return fallbackLink
            }
            guard let tocProgression = tocLocator.locations.progression else {
                return fallbackLink
            }
            guard let plainLink = publication.linkWithHREF(tocLocator.href) else {
                return fallbackLink
            }
            guard let chapterClips = clipsByHref[plainLink.url().string] else {
                return fallbackLink
            }
            guard let clip = searchForClipsByProgression(clips: chapterClips, progression: tocProgression) else {
                return fallbackLink
            }

            guard let audioResource = RelativeURL(url: clip.relativeUrl) else {
                return fallbackLink
            }

            let clipResource = publication.resources.first {
                $0.href == audioResource.string
            }

            guard let clipResource = clipResource else {
                return fallbackLink
            }

            let href = "\(audioResource)#t=\(clip.start)"

            return Link(
                href: href,
                mediaType: clipResource.mediaType,
                title: link.title,
                children: children,
            )
        }

        var clips: OrderedDictionary<String, Double> = [:]

        for clip in bookClips {
            guard let audioResource = RelativeURL(url: clip.relativeUrl) else {
                continue
            }
            let duration = clips[audioResource.string] ?? 0.0
            let end = clip.end
            let start = clip.start
            clips[audioResource.string] = duration + end - start
        }

        return try await Manifest(
            metadata: publication.metadata,
            readingOrder: clips.compactMap { key, value in
                guard var link = publication.linkWithHREF(RelativeURL(epubHREF: key)!) else {
                    return nil
                }

                link.duration = value
                return link
            },
            tableOfContents: publication.tableOfContents().get().asyncMap {
                try await buildAudiobookTocLink(link: $0)
            }.compactMap { $0 },
            )

    }

    private func checkIsReadable(publication: Publication) throws {
        guard !publication.isRestricted else {
          if let error = publication.protectionError {
              throw BookError.openFailed(error)
          } else {
              throw BookError.cancelled
          }
        }
    }
}

func searchForClipsByProgression(clips: [OverlayPar], progression: Double) -> OverlayPar? {
    var startIndex = clips.startIndex
    var endIndex = clips.endIndex
    while (startIndex <= endIndex) {
        let midIndex = Int((startIndex + endIndex) / 2)
        let midItem = clips[midIndex]
        let prevIndex = midIndex - 1
        let prevItem = prevIndex < 0 ? nil : clips[prevIndex]
        if progression > (midItem.locator.locations.progression ?? 0.0) {
            startIndex = midIndex + 1
            continue
        }
        if (prevItem != nil && progression < (midItem.locator.locations.progression ?? 0.0)) {
            endIndex = midIndex - 1
            continue
        }
        return midItem
    }
    return nil
}

File: mobile/modules/readium/android/src/main/java/expo/modules/readium/BookService.kt
@file:OptIn(InternalReadiumApi::class)

package expo.modules.readium

import android.net.Uri
import kotlinx.coroutines.runBlocking
import org.readium.r2.shared.ExperimentalReadiumApi
import org.readium.r2.shared.InternalReadiumApi
import org.readium.r2.shared.publication.Href
import org.readium.r2.shared.publication.Link
import org.readium.r2.shared.publication.Locator
import org.readium.r2.shared.publication.Manifest
import org.readium.r2.shared.publication.Publication
import org.readium.r2.shared.publication.indexOfFirstWithHref
import org.readium.r2.shared.publication.services.positions
import org.readium.r2.shared.publication.services.positionsByReadingOrder
import org.readium.r2.shared.util.RelativeUrl
import org.readium.r2.shared.util.Try
import org.readium.r2.shared.util.Url
import org.readium.r2.shared.util.asset.AssetRetriever
import org.readium.r2.shared.util.asset.DefaultArchiveOpener
import org.readium.r2.shared.util.asset.DefaultFormatSniffer
import org.readium.r2.shared.util.data.Container
import org.readium.r2.shared.util.data.Readable
import org.readium.r2.shared.util.data.decodeString
import org.readium.r2.shared.util.data.decodeXml
import org.readium.r2.shared.util.data.readDecodeOrNull
import org.readium.r2.shared.util.file.FileResourceFactory
import org.readium.r2.shared.util.fromEpubHref
import org.readium.r2.shared.util.getOrElse
import org.readium.r2.shared.util.mediatype.MediaType
import org.readium.r2.shared.util.resource.Resource
import org.readium.r2.shared.util.toUri
import org.readium.r2.shared.util.xml.ElementNode
import org.readium.r2.streamer.PublicationOpener
import org.readium.r2.streamer.parser.epub.EpubParser
import java.io.File
import java.net.URL
import java.util.zip.ZipFile

object BookService {

    /** Returns the resource data as an XML Document at the given [url], or null. */
    @OptIn(InternalReadiumApi::class)
    private suspend inline fun Container<Readable>.readDecodeXmlOrNull(
        url: Url,
    ): ElementNode? =
        readDecodeOrNull(url) { it.decodeXml() }

    private val retriever: AssetRetriever = AssetRetriever(
        FileResourceFactory(),
        DefaultArchiveOpener(), DefaultFormatSniffer()
    )

    @OptIn(InternalReadiumApi::class)
    private val opener: PublicationOpener =
        PublicationOpener(
            EpubParser()
        )

    private var publications: MutableMap<String, Publication> = mutableMapOf()
    private var clips: MutableMap<String, List<OverlayPar>> = mutableMapOf()

    fun extractArchive(archiveUrl: URL, extractedUrl: URL) {
        ZipFile(archiveUrl.path).use { zip ->
            zip.entries().asSequence()
                .filterNot { it.isDirectory }
                .forEach { entry ->
                    zip.getInputStream(entry).use { input ->
                        val newFile = File(extractedUrl.path, entry.name)
                        newFile.parentFile?.mkdirs()
                        newFile.outputStream().use { output ->
                            input.copyTo(output)
                        }
                    }
                }
        }
    }

    fun getPublication(bookUuid: String): Publication? {
        return publications[bookUuid]
    }

    fun getResource(bookUuid: String, link: Link): Resource? {
        val publication = getPublication(bookUuid)
            ?: throw Exception("Publication for book $bookUuid is unopened.")
        return publication.get(link)
    }

    suspend fun getPositions(bookUuid: String): List<Locator> {
        val publication = getPublication(bookUuid)
            ?: throw Exception("Publication for book $bookUuid is unopened.")
        return publication.positions()
    }

    @OptIn(InternalReadiumApi::class)
    fun getClip(bookUuid: String, locator: Locator): OverlayPar? {
        val clips = this.clips[bookUuid] ?: return null

        val fragment = try {
            locator.locations.fragments.first()
        } catch (_: Exception) {
            return null
        }

        return clips.first { it.locator.href == locator.href && it.fragmentId == fragment }
    }

    fun getOverlayClips(bookUuid: String): List<OverlayPar> {
        return this.clips[bookUuid] ?: listOf()
    }

    @OptIn(InternalReadiumApi::class)
    private suspend fun locateFromPositions(bookUuid: String, link: Link): Locator {
        val publication = getPublication(bookUuid)
            ?: throw Exception("Publication for book $bookUuid is unopened.")

        val readingOrderIndex = publication.readingOrder.indexOfFirstWithHref(link.url())
            ?: throw Exception("Could not find a locator for href ${link.href} in reading order for book $bookUuid")

        return publication.positionsByReadingOrder()[readingOrderIndex].first()
    }

    @OptIn(InternalReadiumApi::class)
    suspend fun buildFragmentLocator(bookUuid: String, href: Url, fragment: String): Locator {
        val publication = getPublication(bookUuid)
            ?: throw Exception("Publication for book $bookUuid is unopened.")

        val defaultLocator = Locator(
            href = href,
            mediaType = MediaType.XHTML
        )

        val link = publication.linkWithHref(href) ?: return defaultLocator

        val resource = publication.get(link) ?: return defaultLocator
        val htmlContent = resource.readDecodeOrNull { Try.success(it.decodeString()) }?.getOrNull()
            ?: return defaultLocator
        val fragmentRegex = Regex("id=\"${fragment}\"")
        val startOfFragment = fragmentRegex.find(htmlContent)?.range?.start ?: return defaultLocator
        val progression = startOfFragment.toDouble() / htmlContent.length.toDouble()
        val startOfChapterProgression =
            locateFromPositions(bookUuid, link).locations.totalProgression
                ?: return defaultLocator
        val chapterIndex = publication.readingOrder.indexOfFirstWithHref(link.url())
            ?: return defaultLocator
        val nextChapterIndex = chapterIndex + 1
        val startOfNextChapterProgression = nextChapterIndex.let {
            if (it == publication.readingOrder.size) {
                return@let 1.0
            } else {
                val nextChapterLink = publication.readingOrder[nextChapterIndex]
                return@let locateFromPositions(bookUuid, nextChapterLink).locations.totalProgression
            }
        } ?: return defaultLocator
        val totalProgression =
            startOfChapterProgression + (progression * (startOfNextChapterProgression - startOfChapterProgression))

        return Locator(
            href = href,
            mediaType = MediaType.XHTML,
            locations = Locator.Locations(
                fragments = listOf(fragment),
                progression = progression,
                totalProgression = totalProgression
            )
        )
    }

    fun getFragments(bookUuid: String, locator: Locator): List<OverlayPar> {
        val clips = this.clips[bookUuid] ?: return emptyList()
        return clips.filter { it.locator.href == locator.href }
    }

    fun getFragment(bookUuid: String, clipUrl: String, position: Double): OverlayPar? {
        val clips = this.clips[bookUuid] ?: return null

        val clipsInUrl = clips.filter { it.audioResource == clipUrl }

        return searchForClip(clipsInUrl, position)
    }

    fun getPreviousFragment(bookUuid: String, locator: Locator): OverlayPar? {
        val currentFragment = locator.locations.fragments.firstOrNull() ?: return null
        val clips = this.clips[bookUuid] ?: return null
        val currentIndex =
            clips.indexOfFirst { it.locator.href == locator.href && it.fragmentId == currentFragment }
        if (currentIndex == 0) return null
        val previousIndex = currentIndex.dec()
        return clips[previousIndex]
    }

    fun getNextFragment(bookUuid: String, locator: Locator): OverlayPar? {
        val currentFragment = locator.locations.fragments.firstOrNull() ?: return null
        val clips = this.clips[bookUuid] ?: return null
        val currentIndex =
            clips.indexOfFirst { it.locator.href == locator.href && it.fragmentId == currentFragment }
        if (currentIndex == 0) return null
        val nextIndex = currentIndex.inc()
        return clips[nextIndex]
    }

    fun locateLink(bookUuid: String, link: Link): Locator? {
        val publication = getPublication(bookUuid) ?: return null
        return publication.locatorFromLink(link)
    }

    suspend fun openPublication(bookUuid: String, url: URL, clips: List<OverlayPar>?): Publication {
        if (publications.contains(bookUuid)) {
            return publications[bookUuid]!!
        }

        val file = File(url.toURI())

        require(file.exists())

        val container =
            DirectoryContainer(file).getOrElse { throw Exception("Failed to open publication at $url: ${it.message}") }

        val asset = this.retriever.retrieve(container, MediaType.EPUB)
            .getOrElse { throw Exception("Failed to open publication at $url: ${it.message}") }

        val publication =
            opener.open(asset, allowUserInteraction = false, onCreatePublication = { ->
                val builder = this
                runBlocking {
                    val containerUrl = RelativeUrl("META-INF/container.xml") ?: return@runBlocking
                    val containerXml =
                        builder.container.readDecodeXmlOrNull(containerUrl) ?: return@runBlocking

                    val opfPath = containerXml
                        .getFirst("rootfiles", Namespaces.OPC)
                        ?.getFirst("rootfile", Namespaces.OPC)
                        ?.getAttr("full-path")
                        ?: return@runBlocking

                    val opfUrl = Url.fromEpubHref(opfPath) ?: return@runBlocking
                    val opfXmlDocument =
                        builder.container.readDecodeXmlOrNull(opfUrl) ?: return@runBlocking
                    val packageDocument = PackageDocument.parse(opfXmlDocument, opfUrl)
                        ?: return@runBlocking

                    val manifestItems = packageDocument.manifest

                    @Suppress("UNCHECKED_CAST")
                    val itemById = manifestItems
                        .filter { it.id != null }
                        .associateBy(Item::id) as Map<String, Item>
                    val readingOrder = builder.manifest.readingOrder.toMutableList()

                    for (manifestItem in manifestItems) {
                        val mediaOverlayId = manifestItem.mediaOverlay ?: continue
                        val mediaOverlayHref = itemById[mediaOverlayId]?.href ?: continue
                        val manifestItemHref = manifestItem.href
                        val linkIndex =
                            readingOrder.indexOfFirstWithHref(manifestItemHref) ?: continue
                        val link = readingOrder[linkIndex]
                        readingOrder[linkIndex] =
                            link.addProperties(mapOf("mediaOverlay" to mediaOverlayHref))
                    }

                    builder.manifest = builder.manifest.copy(
                        readingOrder = readingOrder
                    )
                }

            })
                .getOrElse { throw Exception("Failed to open publication at $url: ${it.message}") }

        publications[bookUuid] = publication
        if (clips == null) {
            makeMediaOverlays(bookUuid, publication)
        } else {
            this.clips[bookUuid] = clips
        }

        return publication
    }

    fun buildAudiobookManifest(bookUuid: String): Manifest {
        val publication = getPublication(bookUuid)
            ?: throw Exception("Publication for book $bookUuid is unopened.")
        val bookClips = this.clips[bookUuid]
            ?: throw Exception("Book $bookUuid has no media overlays")

        val clipsByHref = bookClips.groupBy { it.locator.href }

        fun buildAudiobookTocLink(link: Link): Link? {
            val children = link.children.mapNotNull { buildAudiobookTocLink(it) }
            val fallbackLink = children.firstOrNull()?.let {
                Link(
                    href = it.href,
                    mediaType = it.mediaType,
                    title = link.title,
                    duration = 0.0,
                    children = children
                )
            }

            val tocLocator = locateLink(bookUuid, link) ?: return fallbackLink
            val tocProgression = tocLocator.locations.progression ?: return fallbackLink
            val plainLink = publication.linkWithHref(tocLocator.href) ?: return fallbackLink
            val chapterClips = clipsByHref[plainLink.url()] ?: return fallbackLink
            val clip =
                searchForClipByProgression(chapterClips, tocProgression) ?: return fallbackLink

            val clipResource =
                publication.resources.first { it.href.toString() == clip.audioResource }

            val linkUrl = RelativeUrl("${clipResource.href}#t=${clip.start}")
                ?: return fallbackLink

            return Link(
                href = Href(linkUrl),
                mediaType = clipResource.mediaType,
                title = link.title,
                children = children
            )
        }

        val clips = emptyMap<String, Double>().toMutableMap()

        for (clip in bookClips) {
            val audioResource = clip.audioResource
            val duration = clips.getOrDefault(clip.audioResource, 0.0)
            val end = clip.end
            val start = clip.start
            clips[audioResource] = duration + end - start
        }

        return Manifest(
            metadata = publication.metadata,
            tableOfContents = publication.tableOfContents.mapNotNull { buildAudiobookTocLink(it) },
            readingOrder = clips.entries.mapNotNull {
                publication.linkWithHref(
                    Url.fromEpubHref(it.key)!!
                )?.copy(duration = it.value)
            }
        )
    }

    @OptIn(InternalReadiumApi::class)
    private suspend fun makeMediaOverlays(bookUuid: String, publication: Publication) {
        val bookClips: MutableList<OverlayPar> = mutableListOf()

        val mediaOverlayLinks = publication.resources.filter { it.mediaType == MediaType.SMIL }

        for (link in mediaOverlayLinks) {
            val smilResource = publication.get(link)
            try {
                val smilXml =
                    smilResource?.readDecodeOrNull { Try.success(it.decodeXml()) }?.getOrNull()
                        ?: continue
                val mediaOverlays = SmilParser.parse(publication, smilXml, link) ?: continue
                bookClips.addAll(mediaOverlays.clips())
            } finally {
                smilResource?.close()
            }
        }

        this.clips[bookUuid] = bookClips
    }
}

fun searchForClipByProgression(clips: List<OverlayPar>, progression: Double): OverlayPar? {
    var startIndex = 0
    var endIndex = clips.size - 1
    while (startIndex <= endIndex) {
        val midIndex = (startIndex + endIndex) / 2
        val midItem = clips[midIndex]
        val prevIndex = midIndex.dec()
        val prevItem = if (prevIndex < 0) null else clips[prevIndex]
        if (progression > (midItem.locator.locations.progression ?: 0.0)) {
            startIndex = midIndex + 1
            continue
        }
        if (prevItem != null && progression < (prevItem.locator.locations.progression ?: 0.0)) {
            endIndex = midIndex - 1
            continue
        }
        return midItem
    }
    return null
}